src/features.c

```diff
@@ -0,0 +1,344 @@
+cleanup_struct_features_free (struct features_info_s **info)
+{
+  if (info == NULL || *info == NULL)
+    return;
+
+  struct features_info_s *ptr = *info;
+
+  // Free oci_version_min if it is not NULL
+  if (ptr->oci_version_min != NULL)
+    {
+      free ((void *) ptr->oci_version_min);
+      ptr->oci_version_min = NULL; // Set to NULL after freeing
+    }
+
+  // Free oci_version_max if it is not NULL
+  if (ptr->oci_version_max != NULL)
+    {
+      free ((void *) ptr->oci_version_max);
+      ptr->oci_version_max = NULL; // Set to NULL after freeing
+    }
+  if (ptr->hooks != NULL)
+    {
+      for (size_t i = 0; ptr->hooks[i] != NULL; i++)
+        {
+          free ((void *) ptr->hooks[i]);
+        }
+      free (ptr->hooks);
+    }
+  if (ptr->mount_options != NULL)
+    {
+      for (size_t i = 0; ptr->mount_options[i] != NULL; i++)
+        {
+          free ((void *) ptr->mount_options[i]);
+        }
+      free (ptr->mount_options);
+    }
+  if (ptr->linux.namespaces != NULL)
+    {
+      for (size_t i = 0; ptr->linux.namespaces[i] != NULL; i++)
+        {
+          free ((void *) ptr->linux.namespaces[i]);
+        }
+      free (ptr->linux.namespaces);
+    }
+
+  if (ptr->linux.capabilities != NULL)
+    {
+      for (size_t i = 0; ptr->linux.capabilities[i] != NULL; i++)
+        {
+          free ((void *) ptr->linux.capabilities[i]);
+        }
+      free (ptr->linux.capabilities);
+    }
+  if (ptr->linux.seccomp.actions != NULL)
+    {
+      for (size_t i = 0; ptr->linux.seccomp.actions[i] != NULL; i++)
+        {
+          free ((void *) ptr->linux.seccomp.actions[i]);
+        }
+      free (ptr->linux.seccomp.actions);
+    }
+  if (ptr->linux.seccomp.operators != NULL)
+    {
+      for (size_t i = 0; ptr->linux.seccomp.operators[i] != NULL; i++)
+        {
+          free ((void *) ptr->linux.seccomp.operators[i]);
+        }
+      free (ptr->linux.seccomp.operators);
+    }
+  if (ptr->linux.seccomp.archs != NULL)
+    {
+      for (size_t i = 0; ptr->linux.seccomp.archs[i] != NULL; i++)
+        {
+          free ((void *) ptr->linux.seccomp.archs[i]);
+        }
+      free (ptr->linux.seccomp.archs);
+    }
+  if (ptr->annotations.io_github_seccomp_libseccomp_version != NULL)
+    {
+      free ((void *) ptr->annotations.io_github_seccomp_libseccomp_version);
+      ptr->annotations.io_github_seccomp_libseccomp_version = NULL;
+    }
+
+  free (ptr);
+  *info = NULL;
+}
+
+int
+crun_command_features (struct crun_global_arguments *global_args, int argc, char **argv, libcrun_error_t *err)
+{
+  int first_arg = 0, ret = 0;
+  libcrun_context_t crun_context = {
+    0,
+  };
+
+  argp_parse (&run_argp, argc, argv, 0, 0, &options);
+
+  ret = init_libcrun_context (&crun_context, argv[first_arg], global_args, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  // Call the function in features.c to gather the feature information
+  cleanup_struct_features struct features_info_s *info = NULL;
```

Comment:
declarations at the beginning of the function, this applies to other variables in the function too